python/pyiceberg/utils/config.py

```diff
@@ -0,0 +1,130 @@
+
+
+class Config:
+    config: RecursiveDict
+
+    def __init__(self):
+        config = self._from_configuration_files()
+        config = _merge_config(config, self._from_environment_variables(config))
+        self.config = FrozenDict(**config)
+
+    @staticmethod
+    def _from_configuration_files() -> RecursiveDict:
+        """Looks up configuration files
+
+        Takes preference over RecursiveDict that's close to the user
+        """
+        config: RecursiveDict = {}
+        for directory in [os.environ.get(HOME), os.environ.get(PYICEBERG_HOME)]:
+            if directory:
+                path = f"{directory.rstrip('/')}/{PYICEBERG_YML}"
+                if os.path.isfile(path):
+                    with open(path, encoding="utf-8") as f:
+                        try:
+                            file_config = yaml.safe_load(f)
+                            file_config_lowercase = _lowercase_dictionary_keys(file_config)
+                            config = _merge_config(config, file_config_lowercase)
+                        except yaml.YAMLError:
+                            logger.exception("Not valid config yaml: %s", path)
+        return config
+
+    @staticmethod
+    def _from_environment_variables(config: RecursiveDict) -> RecursiveDict:
+        """Reads the environment variables, to check if there are any prepended by PYICEBERG__
+
+        Args:
+            config: Existing configuration that's being amended with configuration from environment variables
+
+        Returns:
+            Amended configuration
+        """
+
+        def set_property(_config: RecursiveDict, path: List[str], config_value: str):
+            key = path[0]
+            if len(path) == 1:
+                _config[key] = config_value
+            else:
+                if key not in _config:
+                    _config[key] = {}
+                set_property(_config[key], path[1:], config_value)  # type: ignore
+
+        for env_var, config_value in os.environ.items():
+            # Make it lowercase to make it case-insensitive
+            env_var_lower = env_var.lower()
+            if env_var_lower.startswith(PYICEBERG.lower()):
+                key = env_var_lower[len(PYICEBERG) :]
+                parts = key.split("_")
+                set_property(config, parts, config_value)
+
+        return config
+
+    def get_catalog_config(self, catalog_name: str) -> RecursiveDict:
+        if CATALOG in self.config:
+            catalog_name_lower = catalog_name.lower()
+            catalogs = self.config[CATALOG]
+            assert isinstance(catalogs, dict), "Configuration path catalogs needs to be an object"
+            if catalog_name_lower in catalogs:
+                catalog_conf = catalogs[catalog_name_lower]
+                assert isinstance(catalog_conf, dict), f"Configuration path catalogs.{catalog_name_lower} needs to be an object"
+                return catalog_conf
+        return EMPTY_DICT
```

Comment:
What about returning None? Otherwise how would you know that the catalog was not defined?

Reply:
Yes, good one. I've changed this by returning a `None`. I like to just return an empty `dict`, so we can merge that dict into other ones, without having to do a lot of `None` checks.